woocommerce/woocommerce-ios
# Product Settings: implemented Featured and Catalog Visibility

Resolve #1825 

## Description
In this PR I implemented the possibility to edit in Product Settings if a product is featured, and the catalog visibility. Since this view was a little bit different from the others that we implemented in Products but has some elements in common with other views like the `ListSelectorViewController`, I [proposed](https://a8c.slack.com/archives/C6H8C3G23/p1585748657112200) a solution based on a tableview, with a cell that shows our List Selector VC embedded. 

| View to implement             |  The solution (discarded) |
:-------------------------:|:-------------------------:
![Schermata 2020-04-02 alle 18 51 27](https://user-images.githubusercontent.com/495617/78276588-871eea00-7513-11ea-9f99-aaa38333c25d.png) | ![Schermata 2020-04-01 alle 15 44 41](https://user-images.githubusercontent.com/495617/78276602-8b4b0780-7513-11ea-8f6e-ad32b5011106.png)


Initially, it seemed like a good idea, which would have hidden the logic of the "Catalog Visibility" section behind an already used and tested view controller, but after implementing it, I ran into some customization problems. In particular problems regarding the resize of the tableview inside the view controller, problems in obtaining the correct size of the tableview content before it was shown on the screen, and problems in deleting the tableview header only in this case. After writing several hundred lines, I realized that it was not the ideal solution, as it created too much overload, and made customization too fragile.

My final solution is based on the suggestion of @shiki: create a specific `CatalogVisibilityViewController`, easier to maintain, also because of all the information inside the VC are static.

## Changes
- Updated the `Product` model to handle the encoding of `featured` and `catalogVisibilityKey` parameters.
- New model `ProductCatalogVisibility” which represent a new entity with all the catalog visibility options (4 options).
- Moved the model `ProductSettings` under Networking, since it can be used easily to update a Product.
- Added the `ProductCatalogVisibilityViewController` that shows and manage the static view.
- In `ProductSettingsRows`, now the order of the various struct that makes up the cells is the same as the one displayed.
- Updated the `ProductSettingsViewModel` to handle the new featured and catalog visibility data.
- `ProductFormViewController` now updates the information coming from the product settings using the `ProductSettings` model.
- Updated `MockProduct` to handle `featured` and `catalogVisibility`.
- Updated `ProductSettingsViewModel` tests.
- Now the method `productSettingsUpdated()` inside the `ProductUpdated` take a `ProductSettings` model + updated tests.

## Testing
1. Navigate to a product detail
2. Press the "more" button in the navigation bar
3. Press "Product Settings" -> You are able to see the cell "Catalog Visibility" with the current catalog visibility.
4. Press the "Catalog Visibility" cell, you will be able to see the Featured Product status + 4 statuses about the catalog visibility (one of them checked).
5. Try to change the featured product status and catalog visibility. Press back.
6. Press "done", and update the product. -> Reopening the Catalog Visibility view, you will be able to see the new featured status and the catalog visibility updated.
7. Try again to edit the featured status or catalog visibility, press back, and discard changes. -> Make sure that everything is like before.

## Screenshots
| Product Settings             |  Catalog Visibility |
:-------------------------:|:-------------------------:
![screen1](https://user-images.githubusercontent.com/495617/78278568-83409700-7516-11ea-942f-b2de001e22ba.png) | ![screen2](https://user-images.githubusercontent.com/495617/78278587-8c316880-7516-11ea-8f9c-5e1ecea5f9a8.png)



Update release notes:
